Discriminator's insert attribute is not respected

zabeda opened the pull request:
This pull request is related to issue "Running into problem with table per class mapping with multiple subclasses" #22

Corrected the errors in the mappings for the entity classes.

RogerKratz commented:
I don't really understand the mapping of the duplicate "typeid" columns in ParentEntity, but... To avoid envers to audit both `<discriminator>` and `<many-to-one>` value (the cause of the exception) that both point to same column, add  `[NotAudited]` on `ParentEntity.TypeId` property (you'll then get a crash in gethashcode/equals impl but that you probably can fix in some other way). 

Also, the revision numbers in used in tests are wrong I guess. Revision 1 will contain your first tx where you've written "no rev".

zabeda commented:
I don't think the mappings for typeid are technically duplicates.  The way it is configured, the discriminator element is there to provide instructions how to discriminate between the different subtypes.  The insert attribute modifies the registration to say that the discriminator is part of another mapping.  Later in the file, we include the many-to-one mapping to create a foreign key to the type table. This is why I changed the TypeId property to reference the ClassTypeEntity class.  

I have modified the GetHashCode and Equals method on ParentEntity.  Also updated the version number checks.

Your solution to turn off auditing for the one field might work for us.  The field should never change once set. Is it correct that once NotAudited is set for a property, that changes to the field will not trigger a new revision, but the field and its value will still be stored in other revisions?  Looking at the data, it seems to indicate this would be the case. 

I am not very familiar with the Envers code base, but I am taking a look at the discriminator code to see if I can add the check of the insert attribute.  I think I see where in the AuditMetadataGenerator class that it could be added. I hacked the code to add the attribute to the audit class and the unit tests passed. The part I need to figure out is how to retrieve the insert attribute true/false value from the discriminator element.  Any suggestions on how to dig that out?

zabeda commented:
Think I found a solution. Running all of the test cases to verify change does not cause any new issues and should be posting an update to the pull request shortly.

RogerKratz commented:
> Is it correct that once NotAudited is set for a property, that changes to the field will not trigger a new revision, but the field and its value will still be stored in other revisions?

No, using that attribute will neither trigger a new revision if modified, nor its values will be stored in other revisions.

RogerKratz commented:
I now saw that you have `insert=false` on the `<discriminator>`, it wasn't there in the original PR I guess?
If that is what you want to fix and you want to provide test + fix for this (would be great), can you please try to make the PR "smaller" and only contain what's needed to expose the bug - smaller class hierarchy, fewer tx in setup?  

(Also, we use tabs and not spaces in envers. A bit strange that you get spaces - if you have a "modern" IDE like a late version of VS, Rider och VS code or similar, it should pick up .editorconfig file I think. Or have you overriden this behavior in your IDE?)

zabeda commented:
I am on the spaces side of the Tabs vs Spaces debate, so I do have VS2019 setup to use spaces.  I modified the files that did not conform to use tabs.  I will make an effort to pay more attention to that in the future.

I have pushed an update.  Please let me know if any other additional changes are needed.

RogerKratz commented on Src/NHibernate.Envers.Tests/NHibernate.Envers.Tests.csproj:
```diff
@@ -7,6 +7,7 @@
 
   <ItemGroup>
     <EmbeddedResource Include="**/*.hbm.xml" Exclude="bin/**" />
+    <None Remove="Integration\Inheritance\TablePerClass\Discriminate\Mapping.hbm.xml" />
```
This change look strange?

zabeda replied:
I will fix this. I copied the hbm file from another test, so might have happened when I copied it over.
